scarb/src/compiler/compilers/executable.rs

```diff
@@ -71,3 +82,114 @@ impl Compiler for ExecutableCompiler {
+                .iter()
+                .map(|executable| originating_function_path(db, *executable))
+                .sorted()
+                .collect_vec();
+            let scarb_toml = unit
+                .main_component()
+                .package
+                .manifest_path()
+                .workspace_relative(ws);
+            anyhow::bail!(ambiguous_executable_error_message(
+                executable_names,
+                scarb_toml
+            ));
+        }
+    };
+
+    compile_executable_function_in_prepared_db(db, executable, diagnostics_reporter, config)
+}
+
+fn ambiguous_executable_error_message(executables: Vec<String>, scarb_toml: &Utf8Path) -> String {
+    let executable_names = executables.clone().join("\n    ");
+
+    let manifest = executables
+        .iter()
+        .map(|function| {
+            let name = function
+                .clone()
+                .split("::")
+                .last()
+                .map(ToString::to_string)
+                .unwrap_or_else(|| function.clone());
+            formatdoc! {r#"
+                [[target.executable]]
+                name = "{name}"
+                function = "{function}"
+            "#,
+            }
+        })
+        .join("\n");
+
+    formatdoc! {r#"
+        more than one executable found in the main crate: 
+            {}
+        help: add a separate `executable` target for each of your executable functions
+        -> {scarb_toml}
+        {manifest}
+        "#,
+        executable_names
+    }
+}
+
+/// Search crates identified by `main_crate_ids` for functions annotated with `#[executable]` attribute.
+/// If `executable_path` is provided, only functions with exactly the same path will be returned.
+fn find_executable_functions(
+    db: &RootDatabase,
+    main_crate_ids: Vec<CrateId>,
+    executable_path: Option<&str>,
+) -> Vec<ConcreteFunctionWithBodyId> {
+    let mut executables: Vec<_> = find_executable_function_ids(db, main_crate_ids)
+        .into_iter()
+        .filter_map(|(id, labels)| {
+            labels
+                .into_iter()
+                .any(|l| l == EXECUTABLE_RAW_ATTR)
+                .then_some(id)
+        })
+        .collect();
+
+    if let Some(executable_path) = executable_path {
+        executables
+            .retain(|executable| originating_function_path(db, *executable) == executable_path);
+    };
+    executables
+}
+
+/// Returns the path to the function that the executable is wrapping.
+///
+/// If the executable is not wrapping a function, returns the full path of the executable.
+fn originating_function_path(db: &RootDatabase, wrapper: ConcreteFunctionWithBodyId) -> String {
+    let semantic = wrapper.base_semantic_function(db);
+    let wrapper_name = semantic.name(db);
+    let wrapper_full_path = semantic.full_path(db.upcast());
+    let Some(wrapped_name) = wrapper_name.strip_prefix(EXECUTABLE_PREFIX) else {
+        return wrapper_full_path;
+    };
```

Comment:
```suggestion
    let Some(function_name) = wrapper_name.strip_prefix(EXECUTABLE_PREFIX) else {
        return wrapper_full_path;
    };
```

It's really confusing when we have both `wrapper_name` and `wrapped_name` nearby.